### PR TITLE
add sort_by (closes #530)

### LINKED ISSUE #530: Add 'sort_by' JME built-in, to sort list elements by a given key
I've made an example at https://numbas.mathcentre.ac.uk/question/32877/sort-a-list-of-lists-by-a-particular-key/. See the definition of sort_by in the Extensions tab.

If the index parameter is a string, it could also sort dictionaries by key. A more adventurous version would take an expression to evaluate for each item, in the manner of `map`.

To do this, add a function `sort_by` to `runtime/scripts/jme-builtin.js`, with the function body taken from the question linked above, and otherwise following the definition of [`sort`](https://github.com/numbas/Numbas/blob/master/runtime/scripts/jme-builtins.js#L1135). Remember to change the signature to take a `TNum` or `TString` parameter before the `TList`; for clarity, it's probably worth defining a version for each of these cases. You could ideally also define a custom `typecheck` function which makes sure that each element of the list is of the right type and has the given key.

### PR DESCRIPTION
**Pull request by ManuelBlanc**
- The new function has 3 variants:
  + List:   `sort_by([[1,1], [2,2], [3,3], 0)`
  + Dict:   `sort_by([["k":1], ["k":2], ["k":3], "k")`
  + Thunk:  `sort_by([[1,1], [2,2], [3,3], 4-n, n)`

- An additional localization message has been added to the en-GB locale
  for type mismatch errors.
- The unit tests now include some sort-related tests.
- The `tests/jme-runtime.js` file has been updated with only the new code.

### CODE REVIEW

**Comment by ManuelBlanc** [comment]
Somethings to do/ be checked before this is merged:
- [ ] The [JME reference](http://numbas-editor.readthedocs.io/en/latest/jme-reference.html) should be updated to include `sort_by()`.
- [ ] The function should be more thoroughly tested by including more edge cases.
- [ ] Consider using `canonical_compare()` in all sort-related functions
- [ ] _Minor: Fix wrong comments._
- [ ] Add myself to the CONTRIBUTORS.md file.

File: runtime/scripts/jme-builtins.js [hunk 1]
```diff
@@ -941,22 +941,178 @@ jme.substituteTreeOps.let = function(tree,scope,allowUnbound) {
...
-            else if(math.lt(a.value,b.value))
-                return -1;
-            else
-                return 0;
+        var listCopy = args[0].value.slice();
+        listCopy.sort(sortCompare); // In-place
+        return new TList(listCopy);
+    }
+});
+var sortByHelper = function(list, accessor, extra) {
```
[the remark this answers is not in the document]
Comment (christianp) [reply runtime/scripts/jme-builtins.js]: `accessor` is not quite the right name for the second argument. It might be a function which transforms the element instead of just accessing one of its properties. `keyFn` might be better.

**Comment by christianp** [comment]
Thanks for this. You've clearly gone to a lot of effort to understand the rather byzantine way JME functions are defined!
